arcgis-java-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/skins/UtilityNetworkTraceSkin.java

```diff
@@ -0,0 +1,539 @@
+        newTraceBorderPane.setVisible(true);
+      }
+    });
+
+    // only show warning if there are insufficient starting points
+    insufficientStartingPointsWarningHBox.visibleProperty().bind(insufficientStartingPointsProperty);
+
+    // only show warning if there are above the minimum starting points
+    aboveMinStartingPointsWarningHBox.visibleProperty().bind(aboveMinimumStartingPointsProperty);
+
+    // bind the trace name to the value in the text field
+    traceNameProperty.bind(traceNameTextField.textProperty());
+
+    // only enable the trace button if trace is enabled
+    runTraceButton.disableProperty().bind(enableTraceProperty.not());
+
+    // only show the results tab when there are results and/or a result is in progress
+    noResultsFoundVBox.visibleProperty().bind(Bindings.and(isTraceInProgressProperty.not(), traceResultsProperty.emptyProperty()));
+
+    // keep trace results in sync with data
+    traceResultsProperty.addListener((ListChangeListener<UtilityNetworkTraceOperationResult>) c -> {
+      while (c.next()) {
+        if (traceResultsProperty.isEmpty()) {
+          tabPane.getSelectionModel().select(newTraceTab);
+        }
+        for (UtilityNetworkTraceOperationResult removedResult : c.getRemoved()) {
+          resultsTabPane.getTabs().remove(findTabForResult(removedResult));
+        }
+        for (UtilityNetworkTraceOperationResult addedResult : c.getAddedSubList()) {
+          Platform.runLater(() -> {
+            // waits for result to finish process via async operation
+            var tab = new UtilityNetworkTraceOperationResultView(this, addedResult);
+            resultsTabPane.getTabs().add(tab);
+            resultsTabPane.getSelectionModel().select(tab);
+          });
+        }
+      }
+    });
+
+    // configure the UI for when a trace is running
+    isTraceInProgressProperty.addListener(((observable, oldValue, newValue) -> {
+      if (newValue) {
+        tabPane.getSelectionModel().select(resultsTab);
+        newTraceTab.setDisable(true);
+        resultsVBox.setVisible(false);
+        traceInProgressVBox.setVisible(true);
+      } else {
+        resultsVBox.setVisible(true);
+        traceInProgressVBox.setVisible(false);
+        newTraceTab.setDisable(false);
+      }
+    }));
+
+    // configure the clear results button and only display when there are trace results
+    clearResultsButton.setOnAction(e -> traceResultsProperty.clear());
+    clearResultsButton.visibleProperty().bind(Bindings.isNotEmpty(traceResultsProperty));
+
+    getChildren().add(root);
+  }
+
+  /**
+   * Returns the Tab that is displaying the data for the provided result.
+   *
+   * @return the Tab
+   * @since 100.15.0
+   */
+  private Tab findTabForResult(UtilityNetworkTraceOperationResult result) {
+    return resultsTabPane.getTabs().stream()
+      .filter(c -> ((UtilityNetworkTraceOperationResultView) c).getResult() == result).findFirst().orElse(null);
```

Comment:
I don't think it makes any difference but it appears that all the tabs are `UtilityNetworkTraceOperationResultView` so you could use `map` like this to get a stream of cast objects.
```
return resultsTabPane.getTabs()
  .stream()
  .map(t -> (UtilityNetworkTraceOperationResultView) t)
  .filter(t -> t.getResult() == result).findFirst().orElse(null);
```